File: java/src/main/java/br/com/emendes/problems/JumpGame.java
package br.com.emendes.problems;

/**
 * É dado a você um array de inteiros nums. Você está posicionado no primeiro index do array,
 * e cada elemento do array representa o comprimento máximo do salto naquela distância.<br>
 * Retorne true se você pode alcançar o último index, ou false caso contrário.<br><br>
 * <p>
 * Restrições:<br>
 * 1 <= nums.length <= 10⁴<br>
 * 0 <= nums[i] <= 10⁵
 */
public class JumpGame {

  public boolean canJump(int[] nums) {
    if (nums.length == 1) return true;

    boolean[] cannotJump = new boolean[nums.length];

    return jump(0, nums, cannotJump);
  }

  private boolean jump(int index, int[] nums, boolean[] cannotJump) {
    if (index >= nums.length - 1) return true;
    if (cannotJump[index]) return false;
    if (nums[index] == 0) {
      cannotJump[index] = true;
      return false;
    }

    for (int jumpLength = nums[index]; jumpLength > 0; jumpLength--) {
      boolean canJump = jump(index + jumpLength, nums, cannotJump);
      if (canJump) return true;
    }

    cannotJump[index] = true;
    return false;
  }

}

File: javascript/problems/JumpGame.js
/**
 * Você recebe um array de inteiros nums. Você esta possicionado no primeiro índice do array,
 * e cada elemento do array representa o máximo comprimento do seu pulo a partir daquela possição.
 *
 * Retorne true se você pode alcançar o último índice, ou false caso contrário.
 *
 * Restrições:
 *
 * 1 <= nums.length <= 10⁴
 *
 * 0 <= nums[i] <= 10⁵
 *
 * @param {number[]} nums
 * @return {boolean}
 */
const canJump = function (nums) {
  let reach = 0;
  const end = nums.length - 1;
  for (let index = 0; index < nums.length && index <= reach && reach < end; index++) {
    reach = Math.max(reach, index + nums[index]);
  }

  return reach >= end;
};

const firstSolution = function (nums) {
  if (nums.length === 1) return true;
  let index = nums.length - 2;
  let jump = 1;
  while (index > 0) {
    if (nums[index] >= jump) jump = 1;
    else jump++;
    index--;
  }

  return nums[0] >= jump;
};
